### PR TITLE
Add flag to inform Squawk that .sql file(s) will be wrapped in a transaction

### PR DESCRIPTION
**Pull request by andrewsmith**
This addresses #260 by adding a new CLI option called `--assume-in-transaction` that can be supplied to Squawk to indicate that an external SQL migration tool will wrap any SQL files in a transaction when run. Squawk can then use this information to accurately apply its linting rules. By default, assume_in_transaction is false, continuing the current behavior for Squawk.

This can also be set in the .squawk.toml config file as well with `assume_in_transaction = true`. The CLI flag overrides this, as is the case with other options.

`assume_in_transaction` is now piped through to each lint rule, in a similar manner to `pg_version`. This required updating all rule functions to accept this new parameter, even if it was unused. I think this could be improved by passing a (`RuleOptions` or similarly named) struct to each rule as opposed to individual parameters. I'm happy to make that change in a subsequent PR, but didn't want to intermix that with this change. 

There were some cases where rules should make use of `assume_in_transaction`. I also updated the linting code and wrote tests for each rule to ensure that the behavior is correct.

I'm happy to rework almost any of this, so please give feedback.

### CODE REVIEW

File: cli/src/config.rs [hunk 1]
```diff
@@ -42,6 +42,8 @@ pub struct Config {
     pub excluded_rules: Vec<RuleViolationKind>,
     #[serde(default)]
     pub pg_version: Option<Version>,
+    #[serde(default)]
+    pub assume_in_transaction: Option<bool>,
```
Comment (andrewsmith) [inline cli/src/config.rs:46]: I picked `assume_transaction` initially, but saw `--assume-in-transaction` as a suggestion in #106. That sounded better to me so I switched to it.

It's relatively easy to rename this, so happy to hear an argument for something else.

File: linter/src/rules/utils.rs [hunk 2]
```diff
@@ -4,9 +4,12 @@ use squawk_parser::ast::{
...
 use std::collections::HashSet;
 
-pub fn tables_created_in_transaction(tree: &[RawStmt]) -> HashSet<String> {
+pub fn tables_created_in_transaction(
+    tree: &[RawStmt],
+    assume_in_transaction: bool,
+) -> HashSet<String> {
     let mut created_table_names = HashSet::new();
-    let mut inside_transaction = false;
+    let mut inside_transaction = assume_in_transaction;
```
Comment (sbdchd) [inline linter/src/rules/utils.rs:12]: curious if you had thoughts on how should we handle the logic down below where we say `inside_transaction = false`?

File: linter/src/rules/prefer_robust_stmts.rs [hunk 3]
```diff
@@ -20,9 +20,13 @@ enum Constraint {
...
 #[must_use]
-pub fn prefer_robust_stmts(tree: &[RawStmt], _pg_version: Option<Version>) -> Vec<RuleViolation> {
+pub fn prefer_robust_stmts(
+    tree: &[RawStmt],
+    _pg_version: Option<Version>,
+    assume_in_transaction: bool,
+) -> Vec<RuleViolation> {
     let mut errs = vec![];
-    let mut inside_transaction = false;
+    let mut inside_transaction = assume_in_transaction;
```
Comment (sbdchd) [inline linter/src/rules/prefer_robust_stmts.rs:29]: ditto here with regards to nested transactions -- which aren't valid in postgres right?

maybe we error or something?

File: linter/src/rules/constraint_missing_not_valid.rs [hunk 4]
```diff
@@ -9,9 +9,9 @@ use squawk_parser::ast::{
 
 /// Return list of spans for offending transactions. From the start of BEGIN to
 /// the end of COMMIT.
-fn not_valid_validate_in_transaction(tree: &[RawStmt]) -> Vec<Span> {
+fn not_valid_validate_in_transaction(tree: &[RawStmt], assume_in_transaction: bool) -> Vec<Span> {
     let mut not_valid_names = HashSet::new();
-    let mut in_transaction = false;
```
Comment (sbdchd) [inline linter/src/rules/constraint_missing_not_valid.rs:14]: also curious what you think about how we should handle nested transactions here as well

Comment (sbdchd) [reply linter/src/rules/constraint_missing_not_valid.rs:14]: like:

```
squawk --assume-in-transaction foo.sql
```

where foo.sql is

```sql
select 1;
begin;
select 2;
commit;
```

**Comment by sbdchd** [comment]
Looks great! Thank you for your thoroughness and the great PR description!

I had a couple questions on how we should handle (what is probably an error / edge case) with nested transactions in the "run in transaction" enabled state stuff

I'm guessing right now squawk doesn't really handle it well so it might be okay for us to ignore

wdyt?

File: linter/src/rules/utils.rs (hunk 2 above)
Comment (andrewsmith) [reply linter/src/rules/utils.rs:12]: It's my impression that the following logic doesn't need to change with respect to `assume_in_transaction`. But there is another caveat described below.

With the external transaction, we would start with `inside_transaction = true`. Any subsequent BEGIN statements would not change `inside_transaction`. [PostgreSQL would merely issue a warning for those when executing](https://www.postgresql.org/docs/current/sql-begin.html#id-1.9.3.47.7).

Any COMMIT would then set `inside_transaction = false`, which is also correct because Postgres would commit then and warn "there is no transaction in progress" for any subsequent COMMITs.

So I think in this case, it should correctly identify tables that were created in *a* transaction in the same tree. However, if there were multiple transactions in one `tree`, it seems like we would get a set of created tables that were possibly from a different transaction. And this seems like it would happen regardless of this new `assume_in_transaction` option. Modifying a test example from  require_concurrent_index_creation.rs:

```
BEGIN;
CREATE TABLE "core_foo" (
"id" bigint GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
"tenant_id" bigint NULL
);
COMMIT;

-- More statements/transactions ...

BEGIN;
-- Should this be flagged by the linter for not being CONCURRENTLY created, since the table was created in a different transaction?
CREATE INDEX "core_foo_tenant_id_4d397ef9" ON "core_foo" ("tenant_id");
COMMIT;
```

However, the likelihood of this being an actual production issue seems low to me. My intuition around that is, worst-case, the table is populated between the CREATE TABLE and CREATE INDEX. That leads to a slow CREATE INDEX, holding an ACCESS EXCLUSIVE lock, etc. But because these are in the same file, the deploy step executing them hasn't finished and if migrations are run before the new code is deployed, nothing is querying that new table yet.

Any thoughts on that? Is that understandable?

Comment (sbdchd) [reply linter/src/rules/utils.rs:12]: Yeah that's sounds great, thank you for the thorough examples and explanation!

File: linter/src/rules/constraint_missing_not_valid.rs (hunk 4 above)
Comment (andrewsmith) [reply linter/src/rules/constraint_missing_not_valid.rs:14]: I reviewed this again and found an edge case that I previously didn't recognize that result in the violation being flagged twice when both --assume-in-transaction was used and an explicit COMMIT was present. I fixed that and wrote a new test for that in 0887e80.

Then I thought about it a little more and changed how that error was being attributed in a396271 (see the commit message for details). This passes tests and should work in the same way, but attribute the violation to the VALIDATE CONSTRAINT statement.

File: linter/src/rules/prefer_robust_stmts.rs (hunk 3 above)
Comment (andrewsmith) [reply linter/src/rules/prefer_robust_stmts.rs:29]: I reviewed this again carefully and feel that the behavior is correct in the presence/absence of --assume-in-transaction and also if there are repeated BEGIN/COMMIT statements.

**Review by andrewsmith — COMMENTED** [review]
Thanks for your review, Steve! I've been busy with other work, but have had a chance to circle back to this.

> I had a couple questions on how we should handle (what is probably an error / edge case) with nested transactions in the "run in transaction" enabled state stuff
>
> I'm guessing right now squawk doesn't really handle it well so it might be okay for us to ignore
>
> wdyt?

I see value in adding a new check for that that is `--assume-in-transaction` aware. I'll try to hack that together after this is merged.

**Comment by andrewsmith** [comment]
I think I am done with changes for this, but want to run it against our migrations and review the outstanding violations. I'll check back once done and re-request a review.

**Review by sbdchd — APPROVED** [review]
:shipit:
